File: javascript/programmers/12983.js
function solution(strs, t) {
  // Python
  strs.sort((a, b) => b.length - a.length)
  const execute = (t, cnt) => {
    if (!t) return cnt
    for (let str of strs)
      str == t.slice(0, str.length)
        ? execute(t.slice(str.length), cnt + 1)
        : null
  }
  return execute(t, 0)
}

console.log(solution(["ab", "na", "n", "a", "bn"], "nabnabn"))
// 그리디 하게 DFS
// na => bn => ab => n

File: python/programmers/12983.py
from collections import deque

answer = float('inf')


def solution(strs, t):

    def execute(t, cnt):
        global answer
        if not t:
            answer = min(answer, cnt)
        for str in strs:
            if str == t[0:len(str)]:
                execute(t[len(str):], cnt+1)
    execute(t, 0)
    return -1 if answer == float('inf') else answer


print(solution(	["app", "ap", "p", "l", "e", "ple", "pp"], "apple"))
